homeassistant/components/recorder/__init__.py

```diff
@@ -0,0 +1,358 @@
+        except sqlalchemy.exc.SQLAlchemyError as e:
+            log_error(e, retry_wait=QUERY_RETRY_WAIT, rollback=True)
+    return []
+
+
+def row_to_state(row):
+    """Convert a database row to a state."""
+    try:
+        return State(
+            row.entity_id, row.state,
+            json.loads(row.attributes),
+            dt_util.UTC.localize(row.last_changed),
+            dt_util.UTC.localize(row.last_updated)
+        )
+    except ValueError:
+        # When json.loads fails
+        _LOGGER.exception("Error converting row to state: %s", row)
+        return None
+
+
+def row_to_event(row):
+    """Convert a databse row to an event."""
+    try:
+        return Event(
+            row.event_type,
+            json.loads(row.event_data),
+            EventOrigin(row.origin),
+            dt_util.UTC.localize(row.time_fired)
+        )
+    except ValueError:
+        # When json.loads fails
+        _LOGGER.exception("Error converting row to event: %s", row)
+        return None
+
+
+def run_information(point_in_time=None):
+    """Return information about current run.
+
+    There is also the run that covers point_in_time.
+    """
+    _verify_instance()
+
+    recorder_runs = get_model('RecorderRuns')
+    if point_in_time is None or point_in_time > _INSTANCE.recording_start:
+        return recorder_runs(
+            end=None,
+            start=_INSTANCE.recording_start,
+            closed_incorrect=False)
+
+    return query('RecorderRuns').filter(
+        (recorder_runs.start < point_in_time) &
+        (recorder_runs.end > point_in_time)).first()
+
+
+def setup(hass, config):
+    """Setup the recorder."""
+    # pylint: disable=global-statement
+    # pylint: disable=too-many-locals
+    global _INSTANCE
+    purge_days = config.get(DOMAIN, {}).get(CONF_PURGE_DAYS)
+
+    db_url = config.get(DOMAIN, {}).get(CONF_DB_URL, None)
+    if not db_url:
+        db_url = DEFAULT_URL.format(
+            hass_config_path=hass.config.path(DEFAULT_DB_FILE))
+
+    _INSTANCE = Recorder(hass, purge_days=purge_days, uri=db_url)
+
+    return True
+
+
+def query(model_name, *args):
+    """Helper to return a query handle."""
+    if isinstance(model_name, str):
+        return Session().query(get_model(model_name), *args)
+    return Session().query(model_name, *args)
+
+
+def get_model(model_name):
+    """Get a model class."""
+    from homeassistant.components.recorder import models
+
+    return getattr(models, model_name)
+
+
+def log_error(e, retry_wait=0, rollback=True,
+              message="Error during query: %s"):
+    """Log about SQLAlchemy errors in a sane manner."""
+    import sqlalchemy.exc
+    if not isinstance(e, sqlalchemy.exc.OperationalError):
+        _LOGGER.exception(e)
+    else:
+        _LOGGER.error(message, str(e))
+    if rollback:
+        Session().rollback()
+    if retry_wait:
+        _LOGGER.info("Retrying failed query in %s seconds", QUERY_RETRY_WAIT)
+        time.sleep(retry_wait)
```

Comment:
If this happens while processing an API request, our web server will be unresponsive for this time period.